src/types.ts

```diff
@@ -24,12 +24,69 @@ export interface Tags {
   };
 }
 
-export interface TokenList {
+// Parameters for investing into a protected investment.
+//
+// The function signature is constant, which is why it's an optional property. The `tokens` property defines the
+// set of input tokens that can be used to deposit into this investment opportunity. The items in the function
+// signature correspond to these `tokens`. For example, if we're investing in the 3Crypto pool, the `tokens` array
+// would be [ETH, WBTC, USDT]. When calling `invest` on `address`, amounts[0] would correspond to the amount of ETH
+// to invest, amounts[1] would correspond to the amount of WBTC to invest, etc. An amount of MAX_UINT256 should be
+// used to signal that you want to invest all tokens you have
+//
+// The same concept also applies to `approves`, which defines whether or not to check and execute approvals for the
+// given token. This is an option to save gas. Similar to `amounts`, the number passed in here specifies how much
+// to approve the spending contract for
+export interface InvestScript {
+  readonly address: string;
+  readonly tokens: {
+    readonly input: TokenInfo[]; // tokens required to invest in this opportunity
+    readonly receipt: TokenInfo[]; // receipt tokens received for investing
+    readonly rewards: TokenInfo[]; // rewards tokens received for investing
+  };
+  readonly signature?: 'function invest(uint256[] amounts, uint256[] approve)';
+}
+
+// Parameters for exiting a protected investment.
+//
+// Function signature meaning is analogous to `InvestScript`. The `amounts` and `claimRewards` parameters are uint256
+// where zero means don't withdraw/claim anything and MAX_UINT256 means withdraw/claim all. Note that not all
+// reward tokens will support arbitrary claim amounts. The `recipient` field specifies the address all output
+// tokens and reward tokens should be sent to
+export interface DivestScript {
+  readonly address: string;
+  readonly tokens: {
+    readonly receipt: TokenInfo[]; // receipt tokens the user needs to exit this opportunity
+    readonly output: TokenInfo[]; // tokens the user will receive after exiting this opportunity
+    readonly rewards: TokenInfo[]; // rewards tokens the user may claim when exiting this opportunity
+  };
+  readonly signature?: 'function divest(uint256[] amounts, address recipient, uint256[] claimRewards)';
+}
+
+export interface Data {
+  // TODO
+}
+
+export interface ProtectionMarketInfo extends TokenInfo {
+  readonly underlying: TokenInfo; // if `underlying.address === 0xEeee....EEeE`, underlying is ETH
+  readonly trigger: string; // trigger contract address for protection markets, or zero address for money markets
```

Comment:
Should we add an optional `imageURI` to each`ProtectionMarketInfo` item as well

Reply:
`ProtectionMarketInfo` extends `TokenInfo` which already has a `logoURI` field

Reply:
Oh got it.

Maybe it's conventional in smart contract dev the way you did it, but doesn't it seem a little more clear to model it this way:

`interface ProtectionMarketInfo {
   readonly receipt: TokenInfo;
   readonly underlying: TokenInfo
 }
`

Reply:
Sorry not sure I'm following—are you suggesting pulling the `receipt` token info from within the `InvestScript` and `DivestScript` types up to the root?

Reply:
Now that we have the `investmentOpportunity` parent type, I think that makes sense

Reply:
Like this 
![image](https://user-images.githubusercontent.com/17163988/131746196-9535a25e-fa3b-40c1-a0f3-2fa2eb5eeba1.png)

Reply:
No, sorry for not being clear...

I was suggesting that the receipt token (of the cozy protection market itself) is more like a child of the market. Right now the market extends a token, so the token's fields will be flat on the market itself.

So, was suggesting removing the `extends TokenInfo` and adding an extra token field called `receipt` which has cozy market receipt token fields.

Reply:
Ah, I see. Well the market itself _is_ the token, i.e. the market contract and the token contract are the same contract address. I think introducing an artificial distinction between the market and the token would be confusing

Reply:
Got it makes sense :)